.github/workflows/build.yml

```diff
@@ -0,0 +1,583 @@
+name: 'Building, Linting, and Testing'
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened]
+  release:
+    types:
+    - created
+
+env:
+  MASTER_BRANCH: master
+  BUILD_HARNESS_VERSION: 0.5.5
+  HUGO_CONFIG: deploy.toml
+  HTMLTEST_CONFIG: .htmltest.codefresh.yml
+  #TMPDIR: /codefresh/volume
+  YARN_BUILD_DISABLED: false
+  COMPONENTS_BUILD: true
+
+jobs:
+  build-and-push-artifacts:
+    name: "Build Artifacts"
+    runs-on: ubuntu-latest
+
+    steps:
+    - name: "Checkout source code at current commit"
+      uses: actions/checkout@v2
+    - name: "Prepare tags for Docker image"
+      if: (github.event_name == 'release' && github.event.action == 'created') || github.event.pull_request.head.repo.full_name == github.repository
+      id: prepare
+      run: |
+        COMMIT_SHA="${GITHUB_SHA}"
+        if [[ $GITHUB_REF == refs/tags/* ]]; then
+          VERSION=${GITHUB_REF#refs/tags/}
+        elif [[ $GITHUB_REF == refs/pull/* ]]; then
+          VERSION=pr-${{ github.event.pull_request.number }}-merge
+          COMMIT_SHA=${{ github.event.pull_request.head.sha }}
+        fi
+        TAGS=${{ github.repository }}:sha-${COMMIT_SHA:0:7}
+        SLIM_TAGS=${{ github.repository }}:slim-sha-${COMMIT_SHA:0:7}
+        if [[ -n $VERSION ]]; then
+          TAGS="$TAGS,${{ github.repository }}:${VERSION}"
+          SLIM_TAGS="$SLIM_TAGS,${{ github.repository }}:slim-${VERSION}"
+        fi
+        if [[ $VERSION =~ ^[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}$ ]]; then
+          TAGS="$TAGS,${{ github.repository }}:latest"
+          SLIM_TAGS="$SLIM_TAGS,${{ github.repository }}:slim-latest"
+        fi
+        echo ::set-output name=tags::${TAGS}
+        echo ::set-output name=slim-tags::${SLIM_TAGS}
+        # generate an individual tag as an output
+        IFS=',' read -ra TAG_ARRAY <<< "${TAGS}"
+        echo ::set-output name=first_tag::${TAG_ARRAY[0]}
+    - name: "Set up Docker Buildx"
+      uses: docker/setup-buildx-action@v1
+    - name: "Login to DockerHub"
+      if: (github.event_name == 'release' && github.event.action == 'created') || github.event.pull_request.head.repo.full_name == github.repository
+      uses: docker/login-action@v1
+      with:
+        username: ${{ secrets.DOCKERHUB_USERNAME }}
+        password: ${{ secrets.DOCKERHUB_PASSWORD }}
+    - name: "Build and push full docker image to DockerHub"
+      id: docker_full_build
+      uses: docker/build-push-action@v2
+      with:
+        tags: ${{ steps.prepare.outputs.first_tag }}
+        #tags: ${{ steps.prepare.outputs.tags }}
+        outputs: type=docker,dest=/tmp/image.tar
+            #type=image,push=${{ (github.event_name == 'release' && github.event.action == 'created') || github.event.pull_request.head.repo.full_name == github.repository }}
+    - name: "List Images"
+      run: |
+        ls -lraht /tmp
+        #docker load --input /tmp/image.tar
```

Comment:
Hrm... let's discuss. If we're uploading the image to docker hub and then also attaching it as an artifact to GitHub, we're doing something wrong probably. E.g. we should be doing one or the other. And the only reason to push to dockerhub was to use in subsequent steps.

Reply:
Note, it's possible to literally just do a single step than runs docker shell commands on a VM. E.g.

```
docker build ...
docker run ...
```
 
To avoid all this complexity pushing docker images as artifacts. We don't need to have steps just for the sake of having steps.